### PR TITLE
Implement Iter::drain(&mut self) with feature iterator_helper

### PR DESCRIPTION
**Pull request by dns2utf8**
The doc text may be a little short

### CODE REVIEW

**Comment by leonardo-m** [comment]
Regarding the example:
`xs.iter().for_each(|x| sum += x).drain();`

Isn't it better written:
`xs.iter().for_each(|x| sum += x);`

**Comment by kennytm** [comment]
@leonardo-m The example uses `.map`, not `.for_each` 😄 

@dns2utf8 I don't think there is consensus on the name of this method yet. See #44546. Furthermore, a new method should not be `#[stable]`.

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -1514,6 +1514,24 @@ pub trait Iterator {
...
+    ///
+    /// ```
+    /// let xs = [0, 1, 2, 3];
+    /// let mut sum = 0;
+    ///
+    /// xs.iter().map(|x| sum += x).drain();
+    /// assert_eq!(6, sum);
+    /// ```
+    #[inline]
+    #[stable(feature = "iterator_helper", since = "1.22.0")]
```
Comment (bluss) [inline src/libcore/iter/iterator.rs]: It's a minor thing, but to keep it as predictable as possible, *iter_methodname* or *iterator_methodname* would be a good feature name. Depends on what the method name will be, of course, just change it with the method name.

**Comment by ollie27** [comment]
This method should probably take `self` by value like all of the other consuming methods like `count`, `last`, `for_each`, `collect`, `fold` etc...

File: src/libcore/iter/iterator.rs [hunk 2]
```diff
@@ -1514,6 +1514,24 @@ pub trait Iterator {
...
     }
 
+    /// Completly consumes an iterator. This is a conveniant method for
+    /// functional programming with iterators.
+    ///
+    /// ```
+    /// let xs = [0, 1, 2, 3];
+    /// let mut sum = 0;
+    ///
+    /// xs.iter().map(|x| sum += x).drain();
```
Comment (sfackler) [inline src/libcore/iter/iterator.rs]: This is a more complex way to write `xs.iter().sum()` or even `xs.iter().for_each(|x| sum += x)`.

**Comment by sfackler** [comment]
I'm not sure I understand why this needs to exist when `for_each` is already present.

**Comment by carols10cents** [comment]
@dns2utf8 looks like @sfackler has a question about how this differs from `for_each`?

**Comment by carols10cents** [comment]
Hi @dns2utf8, due to inactivity, I'm going to close this PR. Please feel free to reopen this when you have a chance to get back to it. Thank you for your contribution! ❤️

**Pull request closed without merging**